### PR TITLE
Lucas S. trainee files for light/color sensor

### PR DESCRIPTION
**Pull request by LucasJSch**
Links and joints were created in a '.xacro' file as a first step for creating a simulated sensor.
In the future, the xacro file created may be instead a reutilization of the 'camera_generic' file. (As it can be seen, they have a very similar structure)

robot

### CODE REVIEW

**Comment by LucasJSch** [comment]
# Ready for review

File: ca_description/urdf/create_2.gv [hunk 1]
```diff
@@ -0,0 +1,59 @@
+digraph G {
```
Comment (eborghi10) [inline ca_description/urdf/create_2.gv]: Beware of pushing these files since they're not needed.

Comment (eborghi10) [reply ca_description/urdf/create_2.gv]: The same applies for meshes.

File: ca_description/urdf/create_2.xacro [hunk 2]
```diff
@@ -16,6 +17,7 @@
     <mesh filename="package://ca_description/meshes/create_2.dae" />
   </xacro:create_base>
 
-  <!-- Roomblock -->
+  <!-- Roomblock  -->
```
Comment (eborghi10) [inline ca_description/urdf/create_2.xacro]: Revert this change.

File: ca_description/urdf/create_light_sensor.xacro [hunk 3]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0"?>
+<robot xmlns:xacro="http://ros.org/wiki/xacro">
+
+<xacro:macro name="light_sensor" params="prefix pitch reflect">
+
+    <xacro:include filename="$(find ca_description)/urdf/common_properties.xacro"/>
+    
+    <xacro:property name="name" value="${prefix}_light_sensor"/>
+    <xacro:property name="parent_link" value="base_link"/>
+    <xacro:property name="x" value="0.095"/>
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: I recommend you to pass the whole origin block as a parameter. See [here](http://wiki.ros.org/xacro#Macros) (*origin element).

File: ca_description/urdf/create_light_sensor.xacro [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    <xacro:property name="y" value="0.14"/>
+    <xacro:property name="z" value="0.065"/>
+    <xacro:property name="R" value="0"/>
+    <xacro:property name="P" value="${pitch}"/>
+    <xacro:property name="Y" value="${pi/2}"/>
+
+
+        <joint name="${name}_joint" type="fixed">
+            <origin xyz="${x} ${y*reflect} ${z}" rpy="0 ${P} 0" doc="x es para adelante, y para izquierda, z para arriba"/>
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Make the comment in English.

File: ca_description/urdf/create_light_sensor.xacro [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            <parent link="${parent_link}" />
+            <child link="${name}_link" />
+            <axis xyz="0 1 0" />
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Since the joint is fixed, is this property necessary?

File: ca_description/urdf/create_light_sensor.xacro [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        </joint>
+
+        <link name="${name}_link">
+                <origin xyz="0 0 0" rpy="0 0 0" />
+        <visual>
+            <origin xyz="0 0 0" rpy="0 ${pi/2} 0" />
+            <geometry>
+                <mesh filename="package://ca_description/meshes/last_lens.stl" scale=".0005 .0005 .0005  "/>
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Scale the model in a graphics editor. And don't apply the scale here.

File: ca_description/urdf/create_light_sensor.xacro [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+        <joint name="${name}_joint" type="fixed">
+            <origin xyz="${x} ${y*reflect} ${z}" rpy="0 ${P} 0" doc="x es para adelante, y para izquierda, z para arriba"/>
+            <parent link="${parent_link}" />
+            <child link="${name}_link" />
+            <axis xyz="0 1 0" />
+        </joint>
+
+        <link name="${name}_link">
+                <origin xyz="0 0 0" rpy="0 0 0" />
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Check the default values in the documentation. Is this really necessary?

File: ca_description/urdf/create_light_sensor.xacro [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        <visual>
+            <origin xyz="0 0 0" rpy="0 ${pi/2} 0" />
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Apply this rotation in the mesh using a graphics editor.

File: ca_description/urdf/create_light_sensor.xacro [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            <geometry>
+                <mesh filename="package://ca_description/meshes/last_lens.stl" scale=".0005 .0005 .0005  "/>
+                <box size=".02 .01 .01" doc="'x' es para arriba, 'y' para adelante y 'z' para la derecha"/>
+            </geometry>
+            <material name="blue">
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Are you seeing this material in Gazebo? In case you don't, see the section "Materials: Using proper colors and textures", [here](http://gazebosim.org/tutorials/?tut=ros_urdf).

File: ca_tools/launch/keyboard_teleop.launch [hunk 10]
```diff
@@ -1,3 +1,3 @@
 <launch>
-  <node pkg="ca_tools" type="key_teleop.py" name="keyboard_teleop" output="screen"/>
+  <node pkg="ca_tools" type="key_teleop.py" name="keyboard_teleop" output="screen" ns="create1"/>
```
Comment (eborghi10) [inline ca_tools/launch/keyboard_teleop.launch]: Rebase against `kinetic-devel` to have a fix for this. Also, don't push this kind of files that are not related to the PR.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
Some changes, PTAL.

**Comment by LucasJSch** [comment]
# Ready for review

**Comment by eborghi10** [comment]
> Links and joints were created in a '.xacro' file as a first step for creating a simulated sensor.
> In the future, the xacro file created may be instead a reutilization of the 'camera_generic' file. (As it can be seen, they have a very similar structure)

Could you add a picture with the current status of the PR?

File: ca_description/urdf/create_base.xacro [hunk 11]
```diff
@@ -54,6 +55,16 @@
     <xacro:create_wheel prefix="right" y_offset="${wheel_separation / -2}" wheel_radius="${wheel_radius}" wheel_width="${wheel_width}"/>
     <xacro:caster_wheel/>
 
+    <xacro:light_sensor prefix="left">
+      <origin xyz="0.095 0.14 0.065" rpy="0 ${pi/4} 0"/>
```
Comment (eborghi10) [inline ca_description/urdf/create_base.xacro]: Since you're sending the origin block, do you need the `prefix`?

File: ca_description/urdf/create_base.xacro [hunk 12]
```diff
@@ -54,6 +55,16 @@
...
+      <optical_origin xyz="0.042 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
```
Comment (eborghi10) [inline ca_description/urdf/create_base.xacro]: Check the `optical_origin` block in both cases, does something change?

Comment (eborghi10) [reply ca_description/urdf/create_base.xacro]: Think about it in this way:

If you move your light sensor around, does the position move w.r.t. the sensor itself?

File: ca_description/urdf/create_light_sensor.xacro [hunk 13]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0"?>
+<robot xmlns:xacro="http://ros.org/wiki/xacro">
+
+<xacro:macro name="light_sensor" params="prefix *origin *optical_origin">
+
+    <xacro:include filename="$(find ca_description)/urdf/common_properties.xacro"/>
+    
+    <xacro:property name="name" value="${prefix}_light_sensor"/>
+    <xacro:property name="parent_link" value="base_link"/>
+
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Remove the extra empty line.

File: ca_description/urdf/create_light_sensor.xacro [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+            <parent link="${parent_link}" />
+            <child link="${name}_link" />
+        </joint>
+
+        <link name="${name}_link">
+        <visual>
+            <geometry>
+                <mesh filename="package://ca_description/meshes/last_lens.stl" />
+            </geometry>
+            <material name="blue">
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: I don't think this declaration here is necessary. Please double check.

Comment (eborghi10) [reply ca_description/urdf/create_light_sensor.xacro]: You already declared its color somewhere else.

File: ca_description/urdf/create_light_sensor.xacro (hunk 13 above)
Comment (eborghi10) [reply ca_description/urdf/create_light_sensor.xacro]: Other places too.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
Some minor comments, PTAL. It's looking great!

File: ca_description/urdf/create_base.xacro (hunk 12 above)
Comment (LucasJSch) [reply ca_description/urdf/create_base.xacro]: I thought about this. But I thought also its maybe better to let the user of this model change the optical frame as he wishes without necessity of changing the model xacro file and (possibly) causing some undesired behaviour. Isn't this considered good practice ?

File: ca_description/urdf/create_base.xacro (hunk 11 above)
Comment (LucasJSch) [reply ca_description/urdf/create_base.xacro]: I think the prefix helps in case that you want to see only one of the two frames in Gazebo or RViz

File: ca_description/urdf/create_light_sensor.xacro (hunk 14 above)
Comment (LucasJSch) [reply ca_description/urdf/create_light_sensor.xacro]: If I don't add this line, the color in RViz changes. I mean, the color is not so important in RViz, but I dont quite understand why it happens this way.

**Comment by LucasJSch** [comment]
> > Links and joints were created in a '.xacro' file as a first step for creating a simulated sensor.
> > In the future, the xacro file created may be instead a reutilization of the 'camera_generic' file. (As it can be seen, they have a very similar structure)
> 
> # Could you add a picture with the current status of the PR?

What do you mean by this?

File: ca_description/urdf/create_base.xacro (hunk 12 above)
Comment (eborghi10) [reply ca_description/urdf/create_base.xacro]: Since you're modeling one specific model, the user shouldn't be able to modify that parameter in particular. I mean, you're modeling the LEGO one, not another.

File: ca_description/urdf/create_light_sensor.xacro (hunk 14 above)
Comment (eborghi10) [reply ca_description/urdf/create_light_sensor.xacro]: Oh! That happens because `material` in this case is a ROS parameter and in the other part of the description, it's Gazebo's. So, leave it as it is now.

**Comment by eborghi10** [comment]
> > > Links and joints were created in a '.xacro' file as a first step for creating a simulated sensor.
> > > In the future, the xacro file created may be instead a reutilization of the 'camera_generic' file. (As it can be seen, they have a very similar structure)
> > 
> > 
> > # Could you add a picture with the current status of the PR?
> 
> What do you mean by this?

I don't know why the letter is so big, sorry.

Put a screenshot of the model, how the sensors are positioned and how is the model being shown in RViz.

File: ca_description/urdf/create_base.xacro (hunk 12 above)
Comment (LucasJSch) [reply ca_description/urdf/create_base.xacro]: Ok. Makes sense. I'll change it

**Comment by LucasJSch** [comment]
# Ready for review

**Comment by eborghi10** [comment]
> Links and joints were created in a '.xacro' file as a first step for creating a simulated sensor.
> In the future, the xacro file created may be instead a reutilization of the 'camera_generic' file. (As it can be seen, they have a very similar structure)
> 
> robot_with_light_sensors

That color is ugly, please use another one. More colors [here](http://wiki.ros.org/simulator_gazebo/Tutorials/ListOfMaterials).

File: ca_description/urdf/create_base.xacro [hunk 15]
```diff
@@ -54,6 +55,13 @@
     <xacro:create_wheel prefix="right" y_offset="${wheel_separation / -2}" wheel_radius="${wheel_radius}" wheel_width="${wheel_width}"/>
     <xacro:caster_wheel/>
 
+    <xacro:light_sensor prefix="left">
```
Comment (eborghi10) [inline ca_description/urdf/create_base.xacro]: Call the parameter `prefix_name` or whatever you consider but it was not clear to me the purpose of that parameter.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
One minor fix, PTAL.

**Comment by LucasJSch** [comment]
# Ready for review

**Review by eborghi10 — APPROVED** [review]
LGTM!

**Comment by LucasJSch** [comment]
# Ready for review

File: ca_color_sensor_plugin/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,33 @@
+cmake_minimum_required(VERSION 2.8.3)
```
Comment (eborghi10) [inline ca_color_sensor_plugin/CMakeLists.txt]: Put the plugin in [this package](https://github.com/RoboticaUtnFrba/create_autonomy/tree/kinetic-devel/ca_gazebo).

File: ca_color_sensor_plugin/include/ca_color_sensor_plugin/color_sensor_plugin.h [hunk 17]
```diff
@@ -0,0 +1,56 @@
+#ifndef GAZEBO_ROS_LIGHT_SENSOR_HH
+    #define GAZEBO_ROS_LIGHT_SENSOR_HH
```
Comment (eborghi10) [inline ca_color_sensor_plugin/include/ca_color_sensor_plugin/color_sensor_plugin.h]: Remove the indentation from line 2 until the end of the file.

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+#include <ros/ros.h>
+#include "ca_color_sensor_plugin/color_sensor_plugin.h"
+
+#include "gazebo_plugins/gazebo_ros_camera.h"
+
+#include <string>
+
+#include <gazebo/sensors/Sensor.hh>
+#include <gazebo/sensors/CameraSensor.hh>
+#include <gazebo/sensors/SensorTypes.hh>
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Are all these includes needed?

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+namespace gazebo
+{
+  // Register this plugin with the simulator
+  GZ_REGISTER_SENSOR_PLUGIN(GazeboRosLight)
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Constructor
+  GazeboRosLight::GazeboRosLight():
+  _nh("color_sensor_plugin"),
+  _fov(6),
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Did you check these values with the real sensor?

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+    if (!ros::isInitialized())
+    {
+      ROS_FATAL_STREAM("A ROS node for Gazebo has not been initialized, unable to load plugin. "
+        << "Load the Gazebo system plugin 'libgazebo_ros_api_plugin.so' in the gazebo_ros package)");
+      return;
+    }
+
+    CameraPlugin::Load(_parent, _sdf);
+    // copying from CameraPlugin into GazeboRosCameraUtils
+    this->parentSensor_ = this->parentSensor;
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Where does `this->parentSensor` come from? I recommend you to copy `_parent`.

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+    this->sensor_color_ = _sdf-> Get<std::string>("sensorColor");
+    _sensorPublisher = _nh.advertise<std_msgs::Bool>(this->publish_topic_name_, 1);
+
+    GazeboRosCameraUtils::Load(_parent, _sdf);
+
+  }
+
+  void GazeboRosLight::GetColorRGB()
+  {
+    //ToDo: Make this prettier with a dictionary.
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Please, do it :smile:

Comment (eborghi10) [reply ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Psst, `std::map`.

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Update the controller
+  void GazeboRosLight::OnNewFrame(const unsigned char *_image,
+    unsigned int _width, unsigned int _height, unsigned int _depth,
+    const std::string &_format)
+  {
+    static int seq=0;
+    GetColorRGB();
+    //this->sensor_update_time_ = this->parentSensor_->GetLastUpdateTime();
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Remove.

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Update the controller
+  void GazeboRosLight::OnNewFrame(const unsigned char *_image,
+    unsigned int _width, unsigned int _height, unsigned int _depth,
+    const std::string &_format)
+  {
+    static int seq=0;
+    GetColorRGB();
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: This can be set only once in `Load()`.

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+    else
+    {
+      if ((*this->image_connect_count_) > 0)
+      {
+        common::Time cur_time = this->world_->SimTime();
+        if (cur_time - this->last_update_time_ >= this->update_period_)
+        {
+          this->PutCameraData(_image);
+          this->PublishCameraInfo();
+          this->last_update_time_ = cur_time;
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: This has to be done at the end of the clause.

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+
+          std_msgs::Bool msg;
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Use [this](http://wiki.ros.org/roscpp/Overview/Publishers%20and%20Subscribers#Intraprocess_Publishing).

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+
+          std::vector<double> current_rgb = std::vector<double>(3);
+
+          double goal_color = 0;
+
+          for (int i=0; i<(_height*_width)-2 ; ++i)
```
Comment (eborghi10) [inline ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: If you want to make it better and support other images types, use:

```c++
i+=current_rgb.size()
```
And access with:
```C++
_image[i+0]
_image[i+1]
...
```

File: ca_description/urdf/create_base.xacro [hunk 27]
```diff
@@ -54,6 +55,13 @@
     <xacro:create_wheel prefix="right" y_offset="${wheel_separation / -2}" wheel_radius="${wheel_radius}" wheel_width="${wheel_width}"/>
     <xacro:caster_wheel/>
 
+    <xacro:light_sensor name_prefix="left" sensor_color="Yellow">
```
Comment (eborghi10) [inline ca_description/urdf/create_base.xacro]: Use colors starting with lowercase.

File: ca_description/urdf/create_light_sensor.xacro [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+        <gazebo reference="${name}_link">
+            <material>Gazebo/DarkGrey</material>
+        </gazebo>
+
+        <gazebo reference="${name}_link">
+            <sensor name='camera' type='camera'>
+            <camera name='__default__'>
+                <horizontal_fov>0.5</horizontal_fov>
+                <image>
+                <width>40</width>
```
Comment (eborghi10) [inline ca_description/urdf/create_light_sensor.xacro]: Can't be smaller?

File: ca_gazebo/launch/prueba.launch [hunk 29]
```diff
@@ -0,0 +1,29 @@
+<launch>
+	<arg name="env" default="prueba"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/prueba.launch]: Don't use names in Spanish.

File: ca_gazebo/launch/prueba.launch [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+
+  	<arg name="paused" value="$(optenv PAUSED false)"/>
+	<arg name="gui" value="$(optenv GUI true)"/>
+	<arg name="debug" value="$(optenv DEBUG false)"/>
+	
+	<env name="GAZEBO_MODEL_PATH" value="$(find ca_gazebo)/models"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/prueba.launch]: Replace this.

File: ca_gazebo/launch/prueba.launch [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+	<include file="$(find gazebo_ros)/launch/empty_world.launch">
+		<arg name="debug" value="$(arg debug)" />
+		<arg name="gui" value="$(arg gui)" />
+		<arg name="paused" value="$(arg paused)"/>
+		<arg name="use_sim_time" value="true"/>
+		<arg name="verbose" value="true"/>
+		<arg name="world_name" value="$(find ca_gazebo)/worlds/$(arg env).world"/>
+	</include>
+
+	<!-- Spawn robot/s -->
```
Comment (eborghi10) [inline ca_gazebo/launch/prueba.launch]: Launch files use [this](https://github.com/RoboticaUtnFrba/create_autonomy/blob/kinetic-devel/ca_gazebo/launch/create_house.launch#L10-L12) instead of what you're doing.

**Review by eborghi10 — CHANGES_REQUESTED** [review]

File: ca_gazebo/launch/prueba.launch (hunk 29 above)
Comment (LucasJSch) [reply ca_gazebo/launch/prueba.launch]: This is an unnecessary file. Deleted it.

File: ca_gazebo/launch/prueba.launch (hunk 30 above)
Comment (LucasJSch) [reply ca_gazebo/launch/prueba.launch]: This is an unnecessary file. Deleted it.

File: ca_gazebo/launch/prueba.launch (hunk 31 above)
Comment (LucasJSch) [reply ca_gazebo/launch/prueba.launch]: This is an unnecessary file. Deleted it.

File: ca_description/urdf/create_light_sensor.xacro (hunk 28 above)
Comment (LucasJSch) [reply ca_description/urdf/create_light_sensor.xacro]: I'm not sure how this will affect the controlling of the robot in the future, but I can make this smaller. I suppose it's more realistic if I make it smaller because in real life the sensor has a very small width and height.

File: ca_color_sensor_plugin/src/color_sensor_plugin.cpp (hunk 26 above)
Comment (LucasJSch) [reply ca_color_sensor_plugin/src/color_sensor_plugin.cpp]: Nice one! I'll do it

**Comment by eborghi10** [comment]
I just realized you pushed to the old PR. For the future: create a new PR targetting to this one. And then you can merge it into this.

**Comment by LucasJSch** [comment]
> I just realized you pushed to the old PR. For the future: create a new PR targetting to this one. And then you can merge it into this.

Perfect. This week I'll cherrypick a little for making this.


## Edit: Should I create a new PR with all the new changes, and delete these changes from this PR?
(By new changes I mean the changes made since the approval of this PR with the visual models of the sensors)

File: ca_description/urdf/create_light_sensor.xacro (hunk 28 above)
Comment (LucasJSch) [reply ca_description/urdf/create_light_sensor.xacro]: I tried with smaller values, and it's not working so well. I'll keep it at 40. At least for now.

**Comment by eborghi10** [comment]
clean up this PR a bit as we discussed in today's meeting.

**Comment by LucasJSch** [comment]
# Cleaned this PR. The deleted commits were added to https://github.com/LucasJSch/create_autonomy/pull/1

**Review by eborghi10 — APPROVED** [review]

**Comment by eborghi10** [comment]
Closing it because this was intended to be used as training.
